mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala

```diff
@@ -945,13 +955,139 @@ class BinaryLogisticRegressionSummary private[classification] (
 private class LogisticAggregator(
     private val numFeatures: Int,
     numClasses: Int,
-    fitIntercept: Boolean) extends Serializable {
+    fitIntercept: Boolean,
+    multinomial: Boolean,
+    standardize: Boolean) extends Serializable {
 
   private var weightSum = 0.0
   private var lossSum = 0.0
 
-  private val gradientSumArray =
-    Array.ofDim[Double](if (fitIntercept) numFeatures + 1 else numFeatures)
+  private val totalCoefficientLength = {
+    val cols = if (fitIntercept) numFeatures + 1 else numFeatures
+    val rows = if (multinomial) numClasses else 1
+    rows * cols
+  }
+
+  private val gradientSumArray = Array.ofDim[Double](totalCoefficientLength)
+
+  /** Update gradient and loss using binary loss function. */
+  private def binaryUpdateInPlace(
+      features: Vector,
+      weight: Double,
+      label: Double,
+      coefficients: Array[Double],
+      gradient: Array[Double],
+      featuresStd: Array[Double],
+      numFeaturesPlusIntercept: Int,
+      standardize: Boolean): Unit = {
+    val margin = - {
+      var sum = 0.0
+      features.foreachActive { (index, value) =>
+        if (featuresStd(index) != 0.0 && value != 0.0) {
+          val x = if (standardize) value / featuresStd(index) else value
+          sum += coefficients(index) * x
+        }
+      }
+      sum + {
+        if (fitIntercept) coefficients(numFeaturesPlusIntercept - 1) else 0.0
+      }
+    }
+
+    val multiplier = weight * (1.0 / (1.0 + math.exp(margin)) - label)
+
+    features.foreachActive { (index, value) =>
+      if (featuresStd(index) != 0.0 && value != 0.0) {
+        val x = if (standardize) value / featuresStd(index) else value
+        gradient(index) += multiplier * x
+      }
+    }
+
+    if (fitIntercept) {
+      gradient(numFeaturesPlusIntercept - 1) += multiplier
+    }
+
+    if (label > 0) {
+      // The following is equivalent to log(1 + exp(margin)) but more numerically stable.
+      lossSum += weight * MLUtils.log1pExp(margin)
+    } else {
+      lossSum += weight * (MLUtils.log1pExp(margin) - margin)
+    }
+  }
+
+  /** Update gradient and loss using multinomial loss function. */
+  private def multinomialUpdateInPlace(
+      features: Vector,
+      weight: Double,
+      label: Double,
+      coefficients: Array[Double],
+      gradient: Array[Double],
+      featuresStd: Array[Double],
+      numFeaturesPlusIntercept: Int,
+      standardize: Boolean): Unit = {
+    /*
+      Note: this can still be used when numClasses = 2 for binary
+      logistic regression without pivoting.
```

Comment:
Just give note to users that since in this case, the coefficients of positive and negative are not shared, so the solution will be the same if no regularization. With regularization, the solution will be different.